### PR TITLE
template: trigger change_mode for dynamic secrets on restore

### PR DESCRIPTION
**Pull request by tgross**
Fixes https://github.com/hashicorp/nomad/issues/9491

When a task is restored after a client restart, the template runner will
create a new lease for any dynamic secret (ex. Consul or PKI secrets
engines). But because this lease is being created in the prestart hook, we
don't trigger the `change_mode`.

This changeset uses the the existence of the task handle to detect a
previously running task that's been restored, so that we can trigger the
template `change_mode` if the template is changed, as it will be only with
dynamic secrets.

---

Note to reviewers: because this requires a bit of a refactor of the `onTemplateRendered`  so that it can be used for both first-pass and re-render, this PR is probably best reviewed commit-by-commit. https://github.com/hashicorp/nomad/issues/9491#issuecomment-742060641 and the diagram in the comment that follow it can walk you thru the problem.

### CODE REVIEW

File: client/allocrunner/taskrunner/interfaces/lifecycle.go
Comment (notnoop) [inline client/allocrunner/taskrunner/interfaces/lifecycle.go:24]: Is this true only if the task is running?  If so, would `IsRunning()` be a more appropriate name? I also suspect we need to watch out for check-then-act races (e.g. process killed immediately after check) - I suspect in this context, it's fine.  Might be worth noting the issue in the comment.

File: client/allocrunner/taskrunner/template/template.go
Comment (notnoop) [inline client/allocrunner/taskrunner/template/template.go:387]: 👍 on extracting the function. This commit is best viewed with `?w=1`: https://github.com/hashicorp/nomad/pull/9636/commits/815f01c9a4a52cd20bc6e17d18b70f7861979cfd?w=1 .

File: client/allocrunner/taskrunner/template/template_test.go
Comment (notnoop) [inline client/allocrunner/taskrunner/template/template_test.go:842]: 👍 to testing that process doesn't get restarted if the template doesn't change.  However, I'm unclear how consul-template state management handles that?  I assumed consul-template library kept template rendition in memory to compare with re-renditions. On nomad agent restart, how does CT library refresh the latest state - does it read the template file and refresh its state?

File: client/allocrunner/taskrunner/template/template_test.go
Comment (notnoop) [inline client/allocrunner/taskrunner/template/template_test.go:819]: I'd suggest using `require`.Equal` and its siblings in this test.

**Review by notnoop — APPROVED** [review]
LGTM. I have one question mostly so I understand the low level mechanics of the consul-template library

File: client/allocrunner/taskrunner/template/template_test.go
Comment (tgross) [reply client/allocrunner/taskrunner/template/template_test.go:842]: > On nomad agent restart, how does CT library refresh the latest state - does it read the template file and refresh its state?

Yeah, there are effectively 3 chunks of state that CT has:
* The `source` which gets reloaded on every client restart.
* The Consul/Vault contents, which gets refreshed on every client restart. That's actually the source of our problem here because for dynamic secrets that's a _write_ which creates a new lease and not a _read_. Every time this triggers, the [`RenderEvent.WouldRender`](https://pkg.go.dev/github.com/hashicorp/consul-template@v0.25.1/manager#RenderEvent) is true. And that includes a client restart.
* The `destination`: the CT library then compares the results with the current contents of the destination and the [`RenderEvent.DidRender`](https://pkg.go.dev/github.com/hashicorp/consul-template@v0.25.1/manager#RenderEvent) is set to false if it didn't actually change on-disk.

So what we're doing is differentiating between the `RenderEvent.WouldRender` and `RenderEvent.DidRender` cases to avoid doing extra restarts we don't need.

Also: this is making me realize we should clarify the behavior around changing the `source` out of band; if you do that while Nomad is running, CT doesn't see it. But then if you restart it will see the change.

File: client/allocrunner/taskrunner/interfaces/lifecycle.go
Comment (tgross) [reply client/allocrunner/taskrunner/interfaces/lifecycle.go:24]: Thanks for pointing that out... it's probably worth me double-checking that if the driver handle goes away we can't NPE in the template runner. At that point we're in a non-working state for this task to be sure but it shouldn't blow up the client process.

I went back-and-forth on the name because we only know that there's a driver handle, not that the process is running and I didn't want future callers to be confused by that. But given the TOCTOU concern you've pointed out, I don't think the caller can assume all that much here except a best effort so I'll change the name and add some commentary about that.

File: client/allocrunner/taskrunner/template/template_test.go
Comment (tgross) [reply client/allocrunner/taskrunner/template/template_test.go]: Done in f03c7b5. The rest of the file uses `t.Fatal` still but I've left the rest of the file alone so as not to blow up this changeset with extraneous stuff.

File: client/allocrunner/taskrunner/interfaces/lifecycle.go
Comment (tgross) [reply client/allocrunner/taskrunner/interfaces/lifecycle.go]: Name changed in 2da0b1f

File: client/allocrunner/taskrunner/template/template_test.go
Comment (tgross) [reply client/allocrunner/taskrunner/template/template_test.go]: I'll have a CHANGELOG and upgrade guide docs PR to go with this (holding off until the giant docs replatform change lands). I'll update the docs on `source` in that PR.

**Comment by tgross** [comment]
Move this back into draft as per https://github.com/hashicorp/nomad/issues/9491#issuecomment-745538850

> After a long soak test it looks like the logmon error is a little misleading and the problem really is in the task restart, so I've got to rework the solution I've got in #9636. I haven't quite pinned down the problem yet.

**Comment by tgross** [comment]
Added changelog and upgrade docs.

File: client/allocrunner/taskrunner/interfaces/lifecycle.go
Comment (krishicks) [inline client/allocrunner/taskrunner/interfaces/lifecycle.go:24]:
```suggestion
	// safely. Ideally prestart hooks should be idempotent whenever possible
```

File: client/allocrunner/taskrunner/template/template.go
Comment (krishicks) [inline client/allocrunner/taskrunner/template/template.go:388]: This comment seems oddly placed, perhaps due to the extraction of the function?

**Review by krishicks — COMMENTED** [review]
This code is a bit beyond my ability to review in-depth; there's too much that I'm unfamiliar with.

File: client/allocrunner/taskrunner/template/template.go
Comment (tgross) [reply client/allocrunner/taskrunner/template/template.go]: Good call. Removed in 62bf55f

**Pull request merged**
